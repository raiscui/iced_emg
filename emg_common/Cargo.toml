[package]
name = "emg_common"
version = "0.1.0"
edition = "2021"
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html


[profile.dev]
split-debuginfo = "unpacked"
[lib]
crate-type = ["rlib"]

[features]
default = []
pool = ["im-rc/pool"]
debug = []

[dependencies]
num-traits = "0.2.15"
# bitflags = "2.0.0-rc.1"
nalgebra = "0.31.4"

tracing = "0.1"
dyn_partial_eq = { path = "../../dyn_partial_eq" }

ordered-float = "3.4.0"
derive_more = "0.99"
im-rc = { version = "15.1.0" }
# imbl = { path = "../../imbl" }
# smol_str = "0.1.23"
# smartstring = "1.0.1"
compact_str = "0.6.1"
# tinyvec = { version = "1.5.1", features = ["alloc"] }
smallvec = { version = "1.10.0", features = ["const_generics"] }
# smallvec = { version = "1.8.0" }
better_any = { path = "../../better_any/better_any", features = [
    "nightly",
    "derive",
] }

bitflags = { path = "../../bitflags_export_const" }


[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-timer = { version = "0.2" }
