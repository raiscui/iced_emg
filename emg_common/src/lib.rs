// ─────────────────────────────────────────────────────────────────────────────

mod func;
mod layout;
mod macros;
mod tools;
// ────────────────────────────────────────────────────────────────────────────────
pub mod animation;
pub mod any;
pub mod display;
pub mod keyboard;
pub mod measures;
pub mod mouse;
pub mod time;
pub mod touch;
pub mod window;
// ────────────────────────────────────────────────────────────────────────────────
pub type Pos<T = Precision> = na::Point2<T>;
pub type Precision = f32;

// ─────────────────────────────────────────────────────────────────────────────

pub use crate::SVec::{smallvec, SmallVec};
pub use ::smallvec as SVec;
pub use better_any;
use better_any::Tid;
pub use compact_str as id_str;
pub use compact_str::CompactString as IdStr;
pub use dyn_partial_eq;
// #[macro_use(vector)]
pub extern crate im_rc as im;
pub use im::Vector;
pub use layout::*;
pub use measures::*;
pub use nalgebra as na;
use num_traits::AsPrimitive;

pub use num_traits;
pub use ordered_float::NotNan;
pub use tools::*;
// ────────────────────────────────────────────────────────────────────────────────

use derive_more::{Display, From};
// pub use smol_str::SmolStr as IdStr;
// pub use smartstring::alias::String as SmString;
// use compact_str::CompactString as SmString;

// pub use tinyvec::{tiny_vec, TinyVec};
// ────────────────────────────────────────────────────────────────────────────────
pub trait TypeCheck {
    const TYPE_NAME: TypeName;
    // fn static_type_name() -> TypeName;
}
pub trait TypeCheckObjectSafe {
    fn type_name(&self) -> TypeName;
}
pub trait TypeCheckObjectSafeTid: for<'a> Tid<'a> + TypeCheckObjectSafe {}
impl<T> TypeCheckObjectSafeTid for T where T: for<'a> Tid<'a> + TypeCheckObjectSafe {}

// // use derive_more::Into;
//TODO full this  "-,/" op
#[derive(Display, Clone, Debug, PartialEq, PartialOrd, Eq)]
pub enum CalcOp<T>
where
    T: Clone + std::fmt::Debug + PartialEq + PartialOrd + Eq,
{
    #[display(fmt = "{} + {}", a, b)]
    Add { a: T, b: T },
    #[display(fmt = "{} * {}", a, b)]
    Mul { a: T, b: NotNan<Precision> },
}

impl<T> CalcOp<T>
where
    T: Clone + std::fmt::Debug + PartialEq + PartialOrd + Eq,
{
    pub fn add(a: T, b: T) -> Self {
        Self::Add { a, b }
    }

    pub fn mul(a: T, b: Precision) -> Self {
        Self::Mul {
            a,
            b: NotNan::new(b.as_()).unwrap(),
        }
    }
}

#[derive(Display, Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct TypeName(IdStr);

impl TypeName {
    pub const fn new(name: IdStr) -> Self
// where
    //     T: AsRef<str>,
    {
        Self(name)
    }
}

impl std::ops::Deref for TypeName {
    type Target = IdStr;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
impl<T> From<T> for TypeName
where
    T: Into<IdStr>,
{
    fn from(v: T) -> Self {
        Self(v.into())
    }
}

// impl From<TypeName> for String {
//     fn from(v: TypeName) -> Self {
//         v.0
//     }
// }

//TODO 创建 (css name, GenericSize) 类型
#[derive(Display, Default, Clone, Debug, From, PartialEq, PartialOrd, Eq)]
#[display(fmt = "{}")]
pub enum GenericSize {
    #[display(fmt = "auto")]
    #[from(ignore)]
    Auto,
    // #[display(fmt = "{}", _0)]
    Length(LogicLength),
    // Percentage(Percent),
    #[display(fmt = "initial")]
    #[from(ignore)]
    Initial,
    #[display(fmt = "inherit")]
    #[from(ignore)]
    Inherit,
    // #[display(fmt = "{}", _0)]
    StringValue(String),
    // #[display(fmt = "{}", _0)]
    Calculation(Box<CalcOp<GenericSize>>),
    Parent(TypeName),
    #[default]
    None,
}

impl GenericSize {
    pub fn zero() -> Self {
        Self::Length(px(0))
    }
}

pub fn parent_ty<T>() -> GenericSize
where
    T: TypeCheck,
{
    GenericSize::Parent(T::TYPE_NAME)
}
pub fn parent_str(type_name: &str) -> GenericSize {
    GenericSize::Parent(TypeName::from(type_name))
}

// impl Default for GenericSize {
//     fn default() -> Self {
//         Self::Length(px(0))

//         // Self::Initial
//     }
// }

impl From<f64> for GenericSize {
    fn from(v: f64) -> Self {
        Self::Length(
            ExactLengthSimplex {
                unit: Unit::Empty,
                value: NotNan::new(v.as_()).unwrap(),
            }
            .into(),
        )
    }
}

impl<T> ::core::ops::Mul<T> for GenericSize
where
    T: AsPrimitive<Precision>,
{
    type Output = GenericSize;
    fn mul(self, rhs: T) -> GenericSize {
        Self::Calculation(Box::new(CalcOp::mul(self, rhs.as_())))
    }
}

impl<T> ::core::ops::Add<T> for GenericSize
where
    T: Into<Self>,
{
    type Output = GenericSize;
    fn add(self, rhs: T) -> GenericSize {
        Self::Calculation(Box::new(CalcOp::add(self, rhs.into())))
    }
}

impl GenericSize {
    #[must_use]
    pub fn get_length_value(&self) -> Precision {
        let msg = format!(
            "directly get length value failed, expected Length Px or None struct, v:{:?}",
            self
        );
        self.try_get_length_value().expect(&msg)
    }

    /// # Errors
    ///
    /// Will return `Err` if `self` does not `Length` and `Length`  unit is not px
    pub fn try_get_length_value(&self) -> Result<Precision, &Self> {
        self.as_length()
            .and_then(|l| l.try_get_number().ok())
            .ok_or(self)
    }

    #[must_use]
    pub const fn as_length(&self) -> Option<&LogicLength> {
        if let Self::Length(v) = self {
            Some(v)
        } else {
            None
        }
    }

    /// Returns `true` if the generic size is [`None`].
    ///
    /// [`None`]: GenericSize::None
    #[must_use]
    pub fn is_none(&self) -> bool {
        matches!(self, Self::None)
    }
}

#[cfg(test)]
mod tests {
    use std::rc::Rc;

    use crate::Vector;

    use crate::into_vector;

    struct XX(Rc<dyn Fn() -> u32>);

    impl std::fmt::Debug for XX {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            f.debug_tuple("XX").finish()
        }
    }

    impl PartialEq for XX {
        fn eq(&self, other: &Self) -> bool {
            std::ptr::eq(
                (std::ptr::addr_of!(*self.0)).cast::<u8>(),
                (std::ptr::addr_of!(*other.0)).cast::<u8>(),
            )
        }
    }

    #[test]
    #[allow(clippy::vtable_address_comparisons)]

    fn it_works() {
        assert_eq!(2 + 2, 4);
        let _f: Vector<i32> = into_vector![1, 2, 3];

        let a = XX(Rc::new(|| 2u32));

        let b = XX(a.0.clone());
        let c = XX(Rc::new(|| 2u32));

        assert_eq!(a, b);
        assert_ne!(a, c);
        assert!(Rc::ptr_eq(&a.0, &b.0));
        assert!(!Rc::ptr_eq(&a.0, &c.0));
    }
}
