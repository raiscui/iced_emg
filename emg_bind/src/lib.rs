#![deny(clippy::all)]
#![deny(clippy::pedantic)]
#![warn(clippy::nursery)]
// #![warn(clippy::cargo)]
// #![deny(unsafe_code)]
#![feature(unboxed_closures, fn_traits, thread_local)]
#![feature(specialization)]
#![feature(drain_filter)]
// ────────────────────────────────────────────────────────────────────────────────
#![feature(convert_float_to_int)] //RafEventRecipe:  (timestamp * 1000.).trunc().to_int_unchecked::<u64>()
#![feature(negative_impls)] // NOTE for Gid refresh
#![feature(auto_traits)]
// #![feature(min_specialization)] // NOTE for Gid refresh
#![feature(box_patterns)]
#![feature(trait_upcasting)]
#![feature(iter_collect_into)]
#![feature(associated_type_defaults)]

// #![feature(associated_type_bounds)]
// bumpalo
// #![feature(allocator_api)]
// ────────────────────────────────────────────────────────────────────────────────
mod application;
mod error;
mod result;
mod sandbox;
mod window;

// ────────────────────────────────────────────────────────────────────────────────
pub mod element;
pub mod executor;
pub mod settings;
pub use emg_element::graph_edit;

// ────────────────────────────────────────────────────────────────────────────────
#[cfg(all(feature = "gpu", feature = "wasm32"))]
compile_error!("current no support");

#[cfg(all(feature = "gpu"))]
// pub use emg_piet_gpu as renderer;
pub use emg_vello as renderer;
#[cfg(all(feature = "gpu"))]
pub use emg_winit as runtime;
// ────────────────────────────────────────────────────────────────────────────────

// ────────────────────────────────────────────────────────────────────────────────
#[cfg(target_arch = "wasm32")]
pub use crate::runtime::widget;
#[cfg(target_arch = "wasm32")]
pub use crate::runtime::Settings;
#[cfg(target_arch = "wasm32")]
pub use crate::runtime::{GelType, GraphType};
#[cfg(target_arch = "wasm32")]
pub use emg_web as runtime;
// ────────────────────────────────────────────────────────────────────────────────
pub use application::Application;
pub use emg;
pub use emg_common as common;
pub use emg_common::any;
pub use emg_common::better_any;
pub use emg_common::mouse;
pub use emg_layout as layout;
pub use emg_msg_macro::emg_msg;
pub use emg_orders::Orders;
pub use emg_shaping as shaping;
pub use emg_state as state;
pub use emg_state::topo;
pub use error::Error;
pub use executor::Executor;
pub use renderer::Renderer;
pub use result::Result;
pub use runtime::{futures, Command};
pub use sandbox::Sandbox;
pub use settings::Settings;

// ─────────────────────────────────────────────────────────────────────────────
pub mod trait_prelude {
    pub use crate::Orders;
}
pub mod gtree_macro_prelude {
    pub use crate::{common::mouse::CLICK, element::gtree_macro_prelude::*};
}
pub mod emg_msg_macro_prelude {
    pub use crate::common::{any, better_any};
    pub use crate::emg_msg;
}
// ────────────────────────────────────────────────────────────────────────────────
#[cfg(test)]
mod test1 {
    use emg_common::{mouse, IdStr};

    fn aaaa(x: impl Into<IdStr>) {
        let f = x.into();
        println!("{f:?}");
    }

    #[test]
    fn aa() {
        let n = "xx";
        aaaa(n);
    }
    #[test]
    fn xx() {
        let f = mouse::CLICK;
        println!("{f:?}");
    }
}

// ────────────────────────────────────────────────────────────────────────────────

// pub use sandbox::Sandbox;
// #[cfg(target_arch = "wasm32")]
// pub use runtime::settings;

// mod g_tree_builder_fn_for_node_item;
// mod g_tree_builder_fn_for_node_item_rc;
// mod graph_store;

// ────────────────────────────────────────────────────────────────────────────────
// mod state_store;
// mod topo_store;

const VEC_SMALL: usize = 4;

// ────────────────────────────────────────────────────────────────────────────────
// ────────────────────────────────────────────────────────────────────────────────
#[cfg(all(test, feature = "gpu"))]
mod tests {
    use emg_msg_macro::emg_msg;

    use crate::element::GTreeBuilderElement;
    use crate::emg_msg_macro_prelude::{any, better_any};
    use crate::gtree_macro_prelude::{gtree, Checkbox, GtreeInitCall, IdStr, Layer, Tid};

    #[emg_msg]
    enum M {
        IncrementPressed,
    }
    fn tree_build() -> GTreeBuilderElement<M> {
        gtree! {
            @="a" Layer [
                @="b" Checkbox::new(false,"abcd",|_|M::IncrementPressed) => [
                    ]
            ]
        }
    }

    fn tree_build2() -> GTreeBuilderElement<M> {
        gtree! {
                @="a" Checkbox::new(false,"abcd",|_|M::IncrementPressed) => [
                    ]
        }
    }

    #[test]
    fn tree_build_tests() {
        let a = tree_build();
        #[cfg(feature = "insta")]
        insta::assert_debug_snapshot!("tree-a", a);
        let b = tree_build2();
        #[cfg(feature = "insta")]
        insta::assert_debug_snapshot!("tree-b", b);
    }
}
#[cfg(all(test, target_arch = "wasm32"))]
mod tests {

    #[allow(unused)]
    use crate::{
        common::{
            better_any::{impl_tid, tid, type_id, Tid, TidAble, TidExt},
            IdStr, TypeCheck,
        },
        layout::{add_values::*, css, styles::*, EmgEdgeItem},
        runtime::{node_ref, EventCallback, EventMessage, GElement, GTreeBuilderElement},
        shaping::{EqShaping, Shaper, Shaping, ShapingUse},
        state::{use_state, CloneState, CloneStateAnchor, StateMultiAnchor},
    };

    #[allow(unused)]
    use std::rc::Rc;
    // #[allow(unused)]
    // use GElement::*;

    #[test]
    fn xx() {
        // let f = node_ref();
    }
}
