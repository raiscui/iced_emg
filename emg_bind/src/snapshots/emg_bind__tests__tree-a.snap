---
source: emg_bind/src/lib.rs
assertion_line: 156
expression: a
---
GTreeBuilderElement::GElementTree(
    "a",
    "with Some Edge Vector...size:0",
    GElement::Layer(
        Layer {
            id: "a",
            children: [],
        },
    ),
    [
        GTreeBuilderElement::GElementTree(
            "b",
            "with Some Edge Vector...size:0",
            GElement::Generic_(
                Checkbox {
                    is_checked: StateVar(
                        false,
                    ),
                    label: "abcd",
                    id: None,
                    width: Simplex(
                        ExactLengthSimplex {
                            unit: Px,
                            value: NotNan(
                                0.0,
                            ),
                        },
                    ),
                },
            ),
            [
                GTreeBuilderElement::Event(
                    "b|CLICK",
                    EventNode<Message>(
                        EventMessage(
                            EventIdentify(
                                2,
                                19,
                            ),
                            "EventMessageFn<Message>",
                        ),
                    ),
                ),
            ],
        ),
    ],
)
